### PR TITLE
[RDY] Flags for bionics

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
# What was changed:

- moved bionic fields (`faulty`, `power_source`, `toggled`, `gun_bionic`, `weapon_bionic` and `armor_interface`) to corresponding bionics flags;
- added descriptions of new bionic flags to `flags.json` and `JSON_FLAGS.md`;
- linted `bionics.json`.

### CODE REVIEW

**Comment by ZhilkinSerg** [comment]
Can someone please suggest why it is not compiling on Jenkins/Travis?

I'm building successfully in Visual Studio.

**Comment by ZhilkinSerg** [comment]
jenkins rebuild

**Comment by codemime** [comment]
> which tries to read Boolean value from obsolete node or check flags array for certain flag (error is shown if obsolete node exists)

You probably don't need such a migration. Bionics aren't written back, so you don't need to worry about back compatibility here. Instead you can simply update the format as needed, and then write a short brief in #19376 when this gets merged.

File: data/json/bionics.json [hunk 1]
```diff
@@ -669,26 +570,22 @@
...
     "occupied_bodyparts": [ [ "HEAD", 3 ] ],
-    "toggled": true
+    "flags": [ "BIONIC_TOGGLED" ]
   },
   {
     "id": "bio_metabolics",
     "type": "bionic",
     "name": "Metabolic Interchange",
-    "description": "Your digestive system and power supply are interconnected.  Energy drains will be regularly counteracted by increasing hunger.  Can be toggled on and off at will.",
+    "description": "Your digestive system and power supply are interconnected.  Energy drains will be regularly counteracted by increasing hunger.  Can be BIONIC_TOGGLED on and off at will.",
```
Comment (AMurkin) [inline data/json/bionics.json]: Unintentional replace?

File: data/json/bionics.json [hunk 2]
```diff
@@ -106,44 +86,36 @@
     "type": "bionic",
     "name": "Battery System",
     "capacity": 10,
-    "description": "You have a battery draining attachment, and thus can make use of the energy contained in normal, everyday batteries.  Use 'E' to consume batteries.  Can be toggled on and off at will.",
+    "description": "You have a battery draining attachment, and thus can make use of the energy contained in normal, everyday batteries.  Use 'E' to consume batteries.  Can be BIONIC_TOGGLED on and off at will.",
```
Comment (AMurkin) [inline data/json/bionics.json]: Unintentional replace?

File: src/json.cpp [hunk 3]
```diff
@@ -214,6 +214,21 @@ bool JsonObject::get_bool(const std::string &name, const bool fallback)
...
 }
 
+bool JsonObject::get_bool_or_flag( const std::string &name, const std::string &flag, const bool fallback, const std::string &flags_node )
+{
+    const std::set<std::string> flags = get_tags( flags_node );
+    bool value = fallback;
+    if( has_bool( name ) ) {
+        std::stringstream err;
+        err << "JsonObject contains legacy node `" << name << "`.  Consider replacing it with `" << flag << "` flag in `" << flags_node << "` node.";
+        JsonObject::throw_error( err.str() );
```
Comment (BevapDin) [inline src/json.cpp]: `throw_error` doesn't return normally, writing any code after it is pointless (`value = ...` will *never* be run).

File: src/json.cpp [hunk 4]
```diff
@@ -214,6 +214,21 @@ bool JsonObject::get_bool(const std::string &name, const bool fallback)
     return jsin->get_bool();
 }
 
+bool JsonObject::get_bool_or_flag( const std::string &name, const std::string &flag, const bool fallback, const std::string &flags_node )
+{
+    const std::set<std::string> flags = get_tags( flags_node );
+    bool value = fallback;
+    if( has_bool( name ) ) {
+        std::stringstream err;
+        err << "JsonObject contains legacy node `" << name << "`.  Consider replacing it with `" << flag << "` flag in `" << flags_node << "` node.";
```
Comment (BevapDin) [inline src/json.cpp]: Why a `stringstream` and not just simple string concatenation?

File: src/json.cpp [hunk 5]
```diff
@@ -1,4 +1,5 @@
 #include "json.h"
+#include "debug.h"
```
Comment (BevapDin) [inline src/json.cpp]: This won't work when the `chkjson` target is build, it only uses "json.cpp" and "chkjson.cpp" - the static variables from "debug.cpp" are not available for it.

**Comment by codemime** [comment]
> Can someone please suggest why it is not compiling on Jenkins/Travis?

Jenkins simply won't run PRs which modify `src/json.h`. I guess for security purposes.

File: data/json/bionics.json (hunk 1 above)
Comment (ZhilkinSerg) [reply data/json/bionics.json]: Thanks.

File: data/json/bionics.json (hunk 2 above)
Comment (ZhilkinSerg) [reply data/json/bionics.json]: Thanks.

**Comment by ZhilkinSerg** [comment]
> Jenkins simply won't run PRs which modify src/json.h. I guess for security purposes.

That explains a lot. I guess I will drop `JsonObject::get_bool_or_flag` function and continue without it.

**Comment by ZhilkinSerg** [comment]
> You probably don't need such a migration. Bionics aren't written back, so you don't need to worry about back compatibility here. Instead you can simply update the format as needed, and then write a short brief in #19376 when this gets merged.

I wanted this change to be backward compatible for third-party mods, but it is prpobably not necessary.

**Comment by kevingranade** [comment]
I scooted get_bool_or_flag() over to bionics.cpp.  Good plan to preserve reverse compatability :+1:

**Comment by ZhilkinSerg** [comment]
> I scooted get_bool_or_flag() over to bionics.cpp. Good plan to preserve reverse compatability 👍

Thanks.

Can we move `get_bool_or_flag()` to `json.cpp` somewhere in the future, so we could apply it to JsonObject anywhere in the code? There are probably more places where it will be good to change boolean nodes with flags, so we will need this common helper. Maybe we should create `json_helper.cpp` or something?

**Comment by kevingranade** [comment]
If a PR needs it outside of bionics.cpp, it can move it to a json_helper module or similar.
It's not functionality of the json format, so I don't want it in json.cpp

**Comment by ZhilkinSerg** [comment]
> If a PR needs it outside of bionics.cpp, it can move it to a json_helper module or similar.
It's not functionality of the json format, so I don't want it in json.cpp

That makes sense. Thanks!

**Pull request closed without merging**